ttg/ttg/func.h

```diff
@@ -95,162 +92,252 @@ namespace ttg {
-  void broadcast(const std::tuple<RangesT...>& keylists, valueT &&value, std::tuple<output_terminalsT...> &t) {
-    static_assert(sizeof...(I)+1 == sizeof...(RangesT),
+  template <size_t i, size_t... I, typename... RangesT, typename valueT, typename... output_terminalsT,
+            ttg::Runtime Runtime = ttg::ttg_runtime,
+            typename = std::enable_if_t<(is_output_terminal_v<output_terminalsT> && ...)>>
+  void broadcast(const std::tuple<RangesT...> &keylists, valueT &&value, std::tuple<output_terminalsT...> &t) {
+    static_assert(sizeof...(I) + 1 == sizeof...(RangesT),
                   "Number of selected output terminals must match the number of keylists!");
     detail::value_copy_handler<Runtime> copy_handler;
     detail::broadcast<0, i, I...>(keylists, copy_handler(std::forward<valueT>(value)), t);
   }
 
-  template <size_t i, typename rangeT, typename... output_terminalsT, ttg::Runtime Runtime = ttg::ttg_runtime>
+  template <size_t i, typename rangeT, typename... output_terminalsT, ttg::Runtime Runtime = ttg::ttg_runtime,
+            typename = std::enable_if_t<(is_output_terminal_v<output_terminalsT> && ...)>>
   void broadcastk(const rangeT &keylist, std::tuple<output_terminalsT...> &t) {
     std::get<i>(t).broadcast(keylist);
   }
 
-  template <size_t i, size_t... I, typename ...RangesT, typename... output_terminalsT, ttg::Runtime Runtime = ttg::ttg_runtime>
-  void broadcastk(const std::tuple<RangesT...>& keylists, std::tuple<output_terminalsT...> &t) {
-    static_assert(sizeof...(I)+1 == sizeof...(RangesT),
+  template <size_t i, size_t... I, typename... RangesT, typename... output_terminalsT,
+            ttg::Runtime Runtime = ttg::ttg_runtime,
+            typename = std::enable_if_t<(is_output_terminal_v<output_terminalsT> && ...)>>
+  void broadcastk(const std::tuple<RangesT...> &keylists, std::tuple<output_terminalsT...> &t) {
+    static_assert(sizeof...(I) + 1 == sizeof...(RangesT),
                   "Number of selected output terminals must match the number of keylists!");
     detail::broadcast<0, i, I...>(keylists, t);
   }
 
   template <typename keyT, typename output_terminalT>
-  std::enable_if_t<!meta::is_void_v<keyT>,void>
-  set_size(const keyT &key, const std::size_t size, output_terminalT &t) {
+  std::enable_if_t<!meta::is_void_v<keyT> && is_output_terminal_v<output_terminalT>, void> set_size(
+      const keyT &key, const std::size_t size, output_terminalT &t) {
     t.set_size(key, size);
   }
 
   template <size_t i, typename keyT, typename... output_terminalsT>
-  std::enable_if_t<!meta::is_void_v<keyT>,void>
-  set_size(const keyT &key, const std::size_t size, std::tuple<output_terminalsT...> &t) {
+  std::enable_if_t<!meta::is_void_v<keyT> && (is_output_terminal_v<output_terminalsT> && ...), void> set_size(
+      const keyT &key, const std::size_t size, std::tuple<output_terminalsT...> &t) {
```

Comment:
God, rly? Put the function name at the end of the line...?

Anyway, why not be more specific on the output terminal types to use overload resolution without the manual constraints here by doing something like:

```C
template <size_t i, typename keyT, typename... output_terminal_valuesT>
  void set_size(
      const keyT &key, const std::size_t size, std::tuple<ttg::Out<keyT, output_terminal_valuesT>...> &t) {
```
We should be able to do that for all functions here except for the broadcast, where we can have several key types.

Reply:
yes, correct, I did not complete the conversion